code/modules/mechanics/MechanicMadness.dm

```diff
@@ -1820,6 +1820,195 @@
+		 <br> Buffer mode is [buffer_string].<br>[buffer_desc]")]"
+
+	New()
+		..()
+		SEND_SIGNAL(src,COMSIG_MECHCOMP_ADD_INPUT,"buffer", PROC_REF(buffer))
+		SEND_SIGNAL(src,COMSIG_MECHCOMP_ADD_INPUT,"set buffer mode", PROC_REF(compSetModel))
+		SEND_SIGNAL(src,COMSIG_MECHCOMP_ADD_CONFIG,"Set Delay",PROC_REF(setDelay))
+		SEND_SIGNAL(src,COMSIG_MECHCOMP_ADD_CONFIG,"Set Buffer Mode",PROC_REF(userSetModel))
+		SEND_SIGNAL(src,COMSIG_MECHCOMP_ADD_CONFIG,"Set Buffer Size",PROC_REF(setBufferSize))
+		SEND_SIGNAL(src,COMSIG_MECHCOMP_ALLOW_MANUAL_SIGNAL)
+		SEND_SIGNAL(src,COMSIG_MECHCOMP_ADD_CONFIG,"Toggle Signal Changing",PROC_REF(toggleDefault))
+
+	proc/setModel(var/model)
+		//Uppercase everything so signal input can be case insensitive
+		model = uppertext(model)
+		if(model == "FIFO")
+			buffer_desc = "This mode outputs from oldest to newest, dropping signals when full."
+			buffer_model = FIFO_BUFFER
+		else if(model == "FILO")
+			buffer_desc = "This mode outputs from newest to oldest, dropping signals when full."
+			buffer_model = FILO_BUFFER
+		else if(model == "RING")
+			model = "ring"
+			buffer_desc = "This mode outputs from oldest to newest, overwriting the oldest signal when full."
+			buffer_model = RING_BUFFER
+		else if(model == "RANDOM")
+			model = "random"
+			buffer_desc = "This mode outputs signals randomly, randomly overwriting previous signals when full."
+			buffer_model = RANDOM_BUFFER
+		else
+			return FALSE
+
+		buffer_string = model
+		tooltip_rebuild = 1
+		qdel(buffer)
+		buffer = list()
+		return TRUE
+
+	proc/compSetModel(var/datum/mechanicsMessage/input)
+		if(setModel(input.signal))
+			LIGHT_UP_HOUSING
+
+	proc/userSetModel(obj/item/W as obj, mob/user as mob)
+		var/model = tgui_input_list(user, "Set the buffer mode to what?", "Mode Selector", list("FIFO","FILO","ring","random"), buffer_string)
+		if(!in_interact_range(src, user) || !can_act(user) || isnull(model))
+			return
+		setModel(model)
+		boutput(user, "You set the buffer mode to [model]")
+
+
+
+	//Thanks delay component
+	proc/setDelay(obj/item/W as obj, mob/user as mob)
+		var/inp = tgui_input_number(user, "Enter delay in 10ths of a second:", "Set delay", cooldown_time, 60, 4)
+		if(!in_interact_range(src, user) || !can_act(user) || isnull(inp))
+			return
+		inp = min(inp, 60)
+		inp = max(4, inp)
+		cooldown_time = inp
+		tooltip_rebuild = 1
+		boutput(user, "Set delay to [inp]")
+
+	proc/setBufferSize(obj/item/W as obj, mob/user as mob)
+		var/inp = tgui_input_number(user,"Set size of signal buffer","Buffer size", buffer_size,buffer_max_size,-1)
+		if(!in_interact_range(src, user) || !can_act(user) || isnull(inp))
+			return
+
+		inp = round(inp)
+		inp = min(inp, buffer_max_size)
+		if(inp == 0)
+			boutput(user,"Buffer size can't be zero.")
+			return
+		if(inp < 0)
+			playsound(src.loc, 'sound/machines/buzz-sigh.ogg', 50,1)
+			boutput(user,"How could a buffer be negative? What are you doing here?")
+			return
```

Comment:
I think this may be redundant since the tgui input does validation anyways but, this can just be written to use a clamp statement.
```suggestion
		if (isnull(inp)) return
		inp = round(inp)
		inp = clamp(inp, 1, buffer_max_size)
```